Update guides for Cy 13, where assertions are now commands (#5081)
cypress-io/cypress-documentation: docs/guides/core-concepts/retry-ability.mdx

## Commands, Queries and Assertions

While all methods you chain off of `cy` in your Cypress tests are commands, it's
important to understand the different rules by which they operate. **Queries**
link up, retrying the entire chain together, while **non-queries** only execute

> once. Assertions are a type of query that's specially displayed in the command

Reviewer: Nit: bullet point each type to make this easier to read. 
```suggestion
once. **Assertions** are a unique query such that it ensures an explicit condition is met to continue executing the Command Queue. 
```

Author: Bullet-pointed the types.

> log. For example, there are 7 queries (2 of them assertions) and 2 non-queries

Reviewer: IMO the logic of assertions are in parallel with query commands under the hood, but they act differently, so keeping the concepts separate makes it easier to digest their purposes and overall behaviors.

Author: They don't act differently any more; that's the whole point of this work.

Assertions are 100% queries in all ways - they are implemented as them and they behave as them, because they are them.

in the test below.

```javascript
it('creates an item', () => {
  // Non-query commands only execute once.
  cy.visit('/')

>   // The queries .focused() and .should() link together,

Reviewer: ```suggestion
  // The .focused() query and .should() assertion link together,
```

Author: .should() is a query, and I'm very intentionally treating it as one in the text. That's why I went through these guides in the first place, to make sure that point is built into our docs.

be smarter and expect the application to potentially update. `cy.get()` queries
the application's DOM, finds the elements that match the selector, and then
passes them to `.find('.todo-list li')`. `.find()` locates a new set of
elements, and passes them to `.should()`. `.should()` then asserts on the list
of found elements (in our case, that it has a length of 1).

- ✅ If the assertion passes, then `.should()` finishes successfully.
- 🚨 If the assertion fails, then Cypress will requery the application's DOM
  again - starting from the top of the chain of linked queries. It will look for
  elements that match `.get().find()` again, and re-run the assertion. If the

>   assertion still fails, Cypress continues retrying until the timeout is

Reviewer: ```suggestion
  assertion still fails, Cypress continues retrying the query chain until the timeout is
```

Author: That's already been mentioned twice in the previous two sentences, I'm not sure stating it a third time in a new way is necessary.

```

The test still passes! `cy.get('.todo-list')` passes immediately - the
`todo-list` exists - but `.should('have.length', 1)` shows the spinning

> indicator, meaning Cypress is still .

Reviewer: I think this needs reworded - nots what this is saying

Reviewer: maybe?
```suggestion
`todo-list` exists - but `.should('have.length', 1)` fails, indicate the application is still showing the spinning indicator.
```

  cy.get('.new-todo').type('todo A{enter}')
  cy.get('.new-todo').type('todo B{enter}')

  cy.get('.todo-list li') // query
    .should('have.length', 2) // query
    .and(($li) => {

>       // 2 expect() statements inside of the .and() query

Reviewer: ```suggestion
      // 2 explicit assertions inside of the .and() query
```

Author: Done. Called them 'mocha assertions' rather than 'explicit assertions'.

      expect($li.get(0).textContent, 'first item').to.equal('todo a')
      expect($li.get(1).textContent, 'second item').to.equal('todo B')
    })
})
```

> Because the first expect statement

Reviewer: ```suggestion
If the first expect statement
```

(`expect($li.get(0).textContent, 'first item').to.equal('todo a')`) fails, the

> second statement is never reached. The command fails after timing out, and the

Reviewer: ```suggestion
second statement is never reached. So the `.and()` command fails after timing out, and the
```

Command Log correctly shows that `.get()` and `.should()` passed, but the "first

> item" assertion failed.

Reviewer: I expected this to say something like the explicit assertion will not cause the .and command to retry"...going into the explanation of the command seems like expected behavior (even though I think we just fixed this).

Author: I'm not sure I'm following you here. The `expect()` statement is inside the callback function for `.and()` - it 100% does cause it to retry, and keep retrying until it times out.

## Built-in assertions

> Often a Cypress command has built-in assertions that will cause the command to

Reviewer: ```suggestion
Often a Cypress command has [Implicit Assertions](link) that will cause the command to
```

Author: I'm not sure it makes sense to link back to the "Getting Started" guide here - this is almost the same content, but written this time with a more granular focus on how commands and queries retry. This is already more detailed than the place it would link.

I have renamed the section "Implicit Assertions" rather than "Built-in Assertions" to better match the terminology in the other guide.

/>

Only queries can be retried, but most other commands still have built-in

> assertions. For example, as described in the "Assertions" section of

Reviewer: From a marketing perspective, I do think keeping references to auto-awaiting is good buzz words to keep in here even if implicit assertions are how we accomplish this

Author: Done.

- Is the element behind another element?
- Does the element have the `disabled` attribute?

> Action commands - such as `.click()` - automatically wait until multiple

Reviewer: This is the first mention of action commands, also missing link

Author: Link added. I haven't bolded it though, randomly bolded words are a pet peeve of mine.

> built-in assertions like these pass, and then it will attempt to click just

Reviewer: ```suggestion
built-in assertions like these pass, and then it will only ever be attempted once.
```
mixing example and def of action command

Author: Good catch. Applied as
> and then it will attempt the action once.
to keep active voice.

## Only queries are retried

Any command that isn't a query, such as `.click()`, will not "chain together" to

> form the subject for later commands the way queries do. Cypress will retry any
> queries _leading up to_ a command, and retry any assertions _after_ a command,
> but commands themselves only execute once. After they've executed, nothing

Reviewer: ```suggestion
form the subject for later commands the way queries do. Cypress will retry any
queries _leading up to_ a failed query,
but commands themselves only execute once. After they've executed, nothing
```

[When Can the Test Click?](https://www.cypress.io/blog/2019/01/22/when-can-the-test-click/)
blog post.

> Because of the default assertions built into every command, and action commands

Reviewer: ```suggestion
Because of the implicit assertions built into every command, and action commands
```

The test finishes before the component calls the `click` prop twice, and without
retrying the assertion `expect(onClick).to.be.calledTwice`.

It could also fail if React or another javascript library decided to rerender
the DOM between clicks.

#### <Icon name="check-circle" color="green" /> Correctly waiting for the stub to be called

We recommend aliasing the stub using the [`.as`](/api/commands/as) command and

> using `cy.get('@alias')` to run assertions.

Reviewer: ```suggestion
using `cy.get('@alias')` to re-run a set of queries and assertions.
```

Author: This change doesn't make much sense to me - we're giving advice directly on how to run assertions about stubs.
